### DISCUSSION TITLE
Providing documentation in the auto-completion popup

### DISCUSSION THREAD

**Discussion by ymc9** [eclipse-langium/langium#1339] [Q&A] [answered]
Hi guys,

I'm working on improving the auto-completion experiences of my DSL, and would like to provide richer documentation in the popup window. Like how Prisma does it:

<img width="659" alt="image" src="https://github.com/eclipse-langium/langium/assets/104139426/e79a4119-9443-4dd4-9b5d-bde164e96543">

I'm looking at the code of [`DefaultCompletionProvider`](https://github.com/eclipse-langium/langium/blob/728b57982b782b31b88ea97aa2a1d3cfe57208b0/packages/langium/src/lsp/completion/completion-provider.ts) but couldn't figure out what's the best way to provide the "documentation" field of `CompletionItem`. I see there's a `DocumentationProvider` that I can customize, however the `DefaultCompletionProvider` doesn't seem to use this service. Are these two services supposed to be connected?

Just wanted to get some confirmation before trying things out more deeply. Thanks :pray:!

**Comment by msujew** [comment, marked as answer]
Hey @ymc9,

the `DocumentationProvider` is currently only used by the `DefaultHoverProvider` as part of the default Langium behavior. It might make sense to adjust the default completion provider to make use of that as well.

Anyway, the best solution for you right now would be to extend the `DefaultCompletionProvider` and override the `createReferenceCompletionItem` method. You can then set the `documentation` field on the returned `CompletionValueItem` object. Note that you only have an `AstNodeDescription` in scope of the `createReferenceCompletionItem` method. You might need to resolve it first. See:

https://github.com/eclipse-langium/langium/blob/728b57982b782b31b88ea97aa2a1d3cfe57208b0/packages/langium/src/references/linker.ts#L203-L210

**Comment by alamarche** [comment]
I basically did what @msujew proposed above in a concept requirements DSL I was working on. I extended the `DefaultCompletionProvider` class and leveraged the `CompletionItem.documentaton` property. Populated it with the same contents that I was passing to the hover provider.

https://github.com/alamarche/reqspec/blob/57acb5089e6d314ca2f8632d26d2290422e1db20/src/language/req-spec-completion-provider.ts#L59C24-L59C24

**Comment by ymc9** [comment]
Thank you for the quick response @msujew , as always! I'll go ahead with the approach then.

@alamarche your sample looks tremendously helpful 🙏 !
